File: backend/app/schemas/media_directory.py
# -*- coding: utf-8 -*-
"""
媒体目录Schema定义
"""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field


class MediaDirectoryBase(BaseModel):
    """媒体目录基础Schema"""
    directory_name: str = Field(..., description="目录名称")
    media_type: Optional[str] = Field(None, description="媒体类型")
    unified_table_name: Optional[str] = Field(None, description="统一资源表名")
    unified_resource_id: Optional[int] = Field(None, description="统一资源ID")
    series_name: Optional[str] = Field(None, description="剧集名称")
    season_number: Optional[int] = Field(None, description="季度编号")


class MediaDirectoryCreate(MediaDirectoryBase):
    """创建媒体目录请求"""
    directory_path: str = Field(..., description="目录完整路径")
    parent_id: Optional[int] = Field(None, description="父目录ID")


class MediaDirectoryUpdate(BaseModel):
    """更新媒体目录请求"""
    unified_resource_id: Optional[int] = Field(None, description="统一资源ID")
    series_name: Optional[str] = Field(None, description="剧集名称")
    season_number: Optional[int] = Field(None, description="季度编号")


class MediaDirectoryResponse(MediaDirectoryBase):
    """媒体目录响应"""
    id: int
    directory_path: str
    parent_id: Optional[int]
    episode_count: int = Field(default=0, description="集数统计")
    has_nfo: bool = Field(default=False, description="是否存在NFO")
    nfo_path: Optional[str]
    has_poster: bool = Field(default=False, description="是否存在海报")
    poster_path: Optional[str]
    has_backdrop: bool = Field(default=False, description="是否存在背景图")
    backdrop_path: Optional[str]
    issue_flags: Dict[str, Any] = Field(default_factory=dict, description="问题标记")
    total_files: int = Field(default=0, description="文件总数")
    total_size: int = Field(default=0, description="总大小(字节)")
    scanned_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DirectoryTreeNode(MediaDirectoryResponse):
    """目录树节点(递归结构)"""
    children: List['DirectoryTreeNode'] = Field(default_factory=list, description="子目录列表")
    has_issues: bool = Field(default=False, description="是否有问题")
    subdir_count: int = Field(default=0, description="子目录数量")

    class Config:
        from_attributes = True


class DirectoryStatistics(BaseModel):
    """目录统计信息"""
    total_files: int = Field(..., description="总文件数")
    total_size: int = Field(..., description="总大小(字节)")
    video_files: int = Field(..., description="视频文件数")
    identified_files: int = Field(..., description="已识别文件数")


class DirectoryDetailResponse(BaseModel):
    """目录详情响应"""
    directory: MediaDirectoryResponse
    statistics: DirectoryStatistics
    files: List[Any] = Field(default_factory=list, description="文件列表")
    nfo_data: Optional[Dict[str, Any]] = Field(None, description="NFO解析数据")


class DirectoryTreeQuery(BaseModel):
    """目录树查询参数"""
    media_type: Optional[str] = Field(None, description="媒体类型筛选")
    parent_id: Optional[int] = Field(None, description="父目录ID")
    load_children: bool = Field(False, description="是否预加载子目录")


class SyncFromFilesRequest(BaseModel):
    """从文件同步目录树请求"""
    base_directory: str = Field(..., description="基础目录路径")


class SyncFromFilesResponse(BaseModel):
    """从文件同步目录树响应"""
    created: int = Field(..., description="新建目录数")
    updated: int = Field(..., description="更新目录数")


class DetectIssuesRequest(BaseModel):
    """检测问题请求"""
    directory_id: Optional[int] = Field(None, description="目录ID(None表示检测所有)")
    media_type: Optional[str] = Field(None, description="媒体类型(None表示所有类型)")


class DetectIssuesResponse(BaseModel):
    """检测问题响应"""
    issues: Dict[str, int] = Field(..., description="问题统计")
    total_issues: int = Field(..., description="总问题数")

File: frontend/src/api/mediaDirectory.ts
/**
 * 媒体目录 API
 */
import request from './request'
import type { TMDBSearchResponse } from '@/types/media'

/**
 * 目录树节点
 */
export interface DirectoryTreeNode {
  id: number
  directory_name: string
  directory_path: string
  parent_id: number | null
  media_type: string | null
  unified_table_name: string | null
  unified_resource_id: number | null
  series_name: string | null
  season_number: number | null
  episode_count: number
  has_nfo: boolean
  nfo_path: string | null
  has_poster: boolean
  poster_path: string | null
  has_backdrop: boolean
  backdrop_path: string | null
  issue_flags: Record<string, any>
  total_files: number
  total_size: number
  scanned_at: string | null
  created_at: string
  updated_at: string
  has_issues: boolean
  children?: DirectoryTreeNode[]
  subdir_count?: number
}

/**
 * NFO 数据
 */
export interface NFOData {
  type?: string
  title?: string
  original_title?: string
  plot?: string
  outline?: string
  tagline?: string
  rating?: number
  year?: number
  runtime?: number
  premiered?: string
  releasedate?: string
  genres?: string[]
  tags?: string[]
  studios?: string[]
  studio?: string
  countries?: string[]
  directors?: string[]
  director?: string
  writers?: string[]
  actors?: Array<{
    name: string
    role?: string
    thumb?: string
  }>
  poster?: string
  fanart?: string
  imdb_id?: string
  tmdb_id?: string
  tvdb_id?: string
  // Episode specific fields
  season?: number
  episode?: number
  aired?: string
  // TV Show specific fields
  status?: string
}

/**
 * 目录详情响应
 */
export interface DirectoryDetailResponse {
  directory: DirectoryTreeNode
  statistics: {
    total_files: number
    total_size: number
    video_files: number
    identified_files: number
  }
  files: any[]
  nfo_data?: NFOData | null
  unified_resource?: {
    id: number
    title: string
    original_title?: string | null
    year?: number | null
    poster_url?: string | null
    backdrop_url?: string | null
    rating?: number | null
    genres?: string[] | null
    overview?: string | null
    media_type?: string
  } | null
}

/**
 * 获取目录树
 */
export function getDirectoryTree(params?: {
  media_type?: string
  parent_id?: number | null
  load_children?: boolean
  issues?: string[]
}) {
  return request.get<DirectoryTreeNode[]>('/media-directories/tree', {
    params: {
      ...params,
      // 数组参数序列化需要处理，这里假设 axios 或 request 库已配置好，或者后端能解析
      // 如果后端 FastAPI 使用 Query(default=None) list 接收，这里直接传数组即可
    }
  })
}

/**
 * 获取目录信息
 */
export function getDirectory(id: number) {
  return request.get<DirectoryTreeNode>(`/media-directories/${id}`)
}

/**
 * 获取目录详情（含文件列表）
 */
export function getDirectoryDetail(id: number) {
  return request.get<DirectoryDetailResponse>(`/media-directories/${id}/detail`)
}

/**
 * 同步目录树（从文件构建）
 */
export function syncDirectories(data: { base_directory: string }) {
  return request.post<{ created: number; updated: number }>('/media-directories/sync', data)
}

/**
 * 检测目录问题
 */
export function detectIssues(data?: { directory_id?: number | null; media_type?: string | null }) {
  return request.post<{ issues: Record<string, number>; total_issues: number }>(
    '/media-directories/detect-issues',
    data || {}
  )
}

/**
 * 关联目录到统一资源（通过TMDB/豆瓣ID 或 直接关联已有资源）
 */
export function linkDirectoryToResource(id: number, data: {
  tmdb_id?: number | null
  douban_id?: string | null
  media_type?: string
  unified_table_name?: string | null
  unified_resource_id?: number | null
}) {
  return request.put<{
    success: boolean
    unified_resource_id: number
    unified_table_name: string
    title: string | null
    message: string
  }>(`/media-directories/${id}/link`, data)
}

/**
 * 搜索TMDB（用于目录识别关联）
 */
export function searchTMDBForDirectory(data: { title: string; year?: number; media_type: string }) {
  return request.post<TMDBSearchResponse>('/media-files/search-tmdb', data)
}

/**
 * 删除目录记录
 */
export function deleteDirectory(id: number) {
  return request.delete(`/media-directories/${id}`)
}
